internal/core/src/segcore/SkipIndex.h

```diff
@@ -0,0 +1,246 @@
+// Copyright (C) 2019-2020 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+
+
+#pragma once
+#include <unordered_map>
+
+#include "common/Types.h"
+#include "log/Log.h"
+
+namespace milvus{
+union MinValue {
+    int8_t int8Value;
+    int16_t int16Value;
+    int32_t int32Value;
+    int64_t int64Value;
+    float floatValue;
+    double doubleValue;
+};
+
+union MaxValue {
+    int8_t int8Value;
+    int16_t int16Value;
+    int32_t int32Value;
+    int64_t int64Value;
+    float floatValue;
+    double doubleValue;
+};
+
+struct FieldChunkMetrics {
+    MinValue min;
+    MaxValue max;
+    bool hasValue_;
+
+    FieldChunkMetrics() : hasValue_(false){};
+};
+
+class SkipIndex{
+public:
+    SkipIndex(){
+        fieldChunkMetrics_ = std::unordered_map<FieldId, std::unordered_map<int64_t, FieldChunkMetrics>>();
+    }
+
+    template <typename T>
+    bool
+    CanSkipUnaryRange(FieldId field_id, int64_t chunk_id, DataType data_type, OpType op_type, T val) const{
+        auto& field_chunk_metrics = GetFieldChunkMetrics(field_id, chunk_id);
+        if(MinMaxUnaryFilter<T>(field_chunk_metrics, data_type, op_type, val)) {
+            return true;
+        }
+        //further more filters for skip, like ngram filter, bf and so on
+        return false;
+    }
+
+    template <typename T>
+    bool
+    CanSkipBinaryRange(FieldId field_id, int64_t chunk_id, DataType data_type,
+            T lower_val, T upper_val, bool lower_inclusive, bool upper_inclusive) const{
+        auto& field_chunk_metrics = GetFieldChunkMetrics(field_id, chunk_id);
+        if(MinMaxBinaryFilter<T>(field_chunk_metrics, data_type, lower_val, upper_val,
+                                 lower_inclusive, upper_inclusive)){
+            return true;
+        }
+        return false;
+    }
+
+    void
+    Load(milvus::FieldId field_id,
+         int64_t chunk_id,
+         milvus::DataType data_type,
+         const void *chunk_data,
+         int64_t count);
+
+private:
+    const FieldChunkMetrics&
+    GetFieldChunkMetrics(FieldId field_id, int chunk_id) const;
+
```

Comment:
one C++ question, it's weird that when using std::is_integral<T> and std::floating_point to replace all these std::is_same<T, ***>, the function cannot be instantiated correctly and complition failed.